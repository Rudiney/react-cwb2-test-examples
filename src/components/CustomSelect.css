.custom-select button {
  padding: 10px;
}

.custom-select ul {
  margin-top: 3;
}

.custom-select ul li {
  list-style-type: none;
}
