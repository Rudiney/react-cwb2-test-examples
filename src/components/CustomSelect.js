import React, { Component } from 'react'
import './CustomSelect.css'

export default class CustomSelect extends Component {
  constructor(p) {
    super(p)
    this.state = { isOpen: false }
    this.toggle = this.toggle.bind(this)
    this.open = this.open.bind(this)
    this.close = this.close.bind(this)
  }

  open() {
    this.props.onOpen && this.props.onOpen()
    this.setState(() => ({ isOpen: true }))
  }

  close() {
    this.setState(() => ({ isOpen: false }))
  }

  toggle() {
    this.state.isOpen ? this.close() : this.open()
  }

  renderOptionsList() {
    if (Object.values(this.props.options).length === 0) {
      return(<ul key='l'><li>Loading...</li></ul>)
    }

    const items = []

    for (const id in this.props.options) {
      items.push(<li key={id} onClick={() => this.props.onChange(id)}>{this.props.options[id]}</li>)
    }

    return (<ul>{items}</ul>)

  }

  render() {
    return (
      <div className='custom-select'>
        <button className='dhdhdhdhddh' onClick={this.toggle}>
          { this.props.value || this.props.placeholder }
        </button>
        { this.state.isOpen && this.renderOptionsList() }
      </div>
    )
  }
}
