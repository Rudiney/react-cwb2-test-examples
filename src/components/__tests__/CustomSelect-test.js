import React from 'react'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'

import CustomSelect from '../CustomSelect.js'

describe('no value and a placeholder', function() {
  it('renders the placeholder', () => {
    const tree = renderer.create(<CustomSelect value={undefined} placeholder='select...'/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('with a value', function() {
  it('display the value in the button', () => {
    const tree = renderer.create(<CustomSelect value='I am selected' placeholder='select...'/>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('the dropdown', function() {
  const options = {
    1: 'One',
    2: 'Two',
    3: 'Three',
  }

  it('clicks to toggle', () => {
    const wrapper = shallow(<CustomSelect options={options} />)
    expect(wrapper.state().isOpen).toBe(false)
    wrapper.find('button').first().simulate('click')
    expect(wrapper.state().isOpen).toBe(true)
    wrapper.find('button').first().simulate('click')
    expect(wrapper.state().isOpen).toBe(false)
  })

  it('renders a ul>li with the options', function() {
    const wrapper = shallow(<CustomSelect options={options} />)
    wrapper.setState({ isOpen: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('triggers a onChange', function() {
    const onChange = jest.fn()
    const wrapper = shallow(<CustomSelect options={options} onChange={onChange} />)
    wrapper.setState({ isOpen: true })
    expect(onChange).not.toBeCalled()
    wrapper.find('ul li').at(1).simulate('click')
    expect(onChange).toBeCalledWith('2')
  })

  it('triggers a onOpen', function() {
    const onOpen = jest.fn()
    const wrapper = shallow(<CustomSelect options={options} onOpen={onOpen} />)
    wrapper.setState({ isOpen: true })
    wrapper.find('button').first().simulate('click')
    expect(onOpen).not.toBeCalled()
    wrapper.find('button').first().simulate('click')
    expect(onOpen).toBeCalled()
  })
})
