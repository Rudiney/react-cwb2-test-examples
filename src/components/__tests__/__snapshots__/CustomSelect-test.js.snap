exports[`no value and a placeholder renders the placeholder 1`] = `
<div
  className="custom-select">
  <button
    className="dhdhdhdhddh"
    onClick={[Function]}>
    select...
  </button>
</div>
`;

exports[`the dropdown renders a ul>li with the options 1`] = `"<div class=\"custom-select\"><button class=\"dhdhdhdhddh\"></button><ul><li>One</li><li>Two</li><li>Three</li></ul></div>"`;

exports[`with a value display the value in the button 1`] = `
<div
  className="custom-select">
  <button
    className="dhdhdhdhddh"
    onClick={[Function]}>
    I am selected
  </button>
</div>
`;
